doc/tutorial/laminar.md

```diff
@@ -0,0 +1,675 @@
+    ...
+          onInput.mapToValue --> makeDataItemUpdater[String](id, {
+            (item, newLabel) =>
+              item.copy(label = newLabel)
+          }),
+    ...
+  end renderDataItem
+
+  def makeDataItemUpdater[A](id: DataItemID,
+      f: (DataItem, A) => DataItem): Observer[A] =
+    dataVar.updater { (data, newValue) =>
+      data.map { item =>
+        if item.id == id then f(item, newValue) else item
+      }
+    }
+  end makeDataItemUpdater
+{% endhighlight %}
+
+As a final refactoring step, we notice the following pattern:
+
+{% highlight scala %}
+        input(
+          typ := "text",
+          value <-- someSignalOfString,
+          onInput.mapToValue --> someObserverOfString,
+        )
+{% endhighlight %}
+
+It represents an `input` text whose `value` is linked to `Signal[String]`, which can be updated through some `Observer[String]`.
+We can define a separate method for this pattern as follows:
+
+{% highlight scala %}
+  def renderDataItem(id: DataItemID, itemSignal: Signal[DataItem]): Element =
+    ...
+      td(
+        inputForString(
+          itemSignal.map(_.label),
+          makeDataItemUpdater(id, { (item, newLabel) =>
+            item.copy(label = newLabel)
+          })
+        )
+      ),
+    ...
+  end renderDataItem
+
+  def inputForString(valueSignal: Signal[String],
+      valueUpdater: Observer[String]): Input =
+    input(
+      typ := "text",
+      value <-- valueSignal,
+      onInput.mapToValue --> valueUpdater,
+    )
+  end inputForString
+{% endhighlight %}
+
+Consider what this method represents.
+It takes data model values as arguments, and returns a Laminar element manipulating those values.
+This is what many UI frameworks call a *component*.
+In Laminar, components are nothing but methods manipulating time-varying data and returning Laminar elements.
+
+## Editing prices and counts
+
+To finish our application, we should also be able to edit *prices* and *counts*.
+
+For the prices, we start with a "component" method building an `Input` that manipulates `Double` values.
+
+{% highlight scala %}
+  def inputForDouble(valueSignal: Signal[Double],
+      valueUpdater: Observer[Double]): Input =
+    val strValue = Var[String]("")
+    input(
+      typ := "text",
+      value <-- strValue.signal,
+      onInput.mapToValue --> strValue,
+      valueSignal --> strValue.updater[Double] { (prevStr, newValue) =>
+        if prevStr.toDoubleOption.contains(newValue) then prevStr
+        else newValue.toString
+      },
+      strValue.signal --> { valueStr =>
+        valueStr.toDoubleOption.foreach(valueUpdater.onNext)
+      },
+    )
+  end inputForDouble
+{% endhighlight %}
+
+It is more complicated than the one for `String`s because of the need for parsing and formatting.
+This complexity perhaps better highlights the benefit of encapsulating it in a dedicated method (or component).
+
+We leave it to the reader to understand the details of this method.
```

Comment:
The one thing I do not understand even after reading carefully: Why are the bindings between valueSignal / strValue / valueUpdater part of the `input` element?